### ISSUE TITLE
Shaders not loaded properly when RetroArch is started trough command-line (glcore)

### ISSUE THREAD

**Issue by LazyBumHorse** [libretro/RetroArch#8834]
## Description

When a game is loaded via the command-line, shaders won't be loaded, despite showing up as active in the shader menu. Only after pressing 'Apply Changes' will they be loaded.

This seems to only happen with the glcore video driver.

### Steps to reproduce the bug

1. Set video driver to glcore
2. Load a slang shader / Set a shader preset
3. Exit RetroArch
4. Run `retroarch.exe -L .\cores\snes9x_libretro.dll "Your Favorite Game.sfc"` or `retroarch.exe -L .\cores\mednafen_psx_hw_libretro.dll "Your Other Favorite Game.cue"`
5. Observe no shaders
6. Run RetroArch normally
7. Load Game
8. Observe shaders

### Bisect Results

It looks like this only happens with the glcore video driver. When using gl with GLSL shaders or vulkan with Slang shaders, the shaders are loaded properly. I haven't tested D3D or cg shaders, but at least the shader language doesn't seem to be the issue.

### What the logs shows

tl;dr exactly what you get from the description

The only difference in the logs is that, when started normally, this
```
[INFO] [GLCore]: Loading stock shader.
[INFO] [slang]: Building pass #0 (N/A)
(...)
```
appears when in the menu and `[INFO] [slang]: Compiling shader (...)` appears immediately after the game is loaded.
When started via the command-line, only the first message appears. It is not until 'Apply Changes' is pressed that the second message appears.

### Version/Commit

- RetroArch: 1.7.7 / 8092f601a6

### Environment information

- Windows 7 x64
- Compiler: GCC 7.3.0

**Comment by KrokusPokus** [comment]
Additional info

This also seems to be depend on the the core or system. I can confirm slang shaders not loading when launching a game from command line when using these cores:
bsnes_balanced_libretro
bsnes_mercury_balanced_libretro
mednafen_psx_libretro
nestopia_libretro
scummvm_libretro

But shaders seem to be loaded/compiled properly when using these cores:
genesis_plus_gx_libretro
mame_libretro
mame2015_libretro

**Comment by LazyBumHorse** [comment]
Currently debugging some.
I noticed that in glcore, this check for slang support fails:
https://github.com/libretro/RetroArch/blob/5c7a5fdba0120566519bc85e42640fb2804256bb/gfx/video_shader_parse.c#L1204
This function tests `video_driver_get_flags()` and `video_context_driver_get_flags()`.
At this point only `video_context_driver_get_flags()` works and the flags returned for me are 129, so the slang shader flag (2^11 = 2048) is not set, which is why it falls back to the stock shader.

Here's the log (with some custom DEBUG lines) starting at the above function call
> [DEBUG] video_context_driver_get_flags() = 129
> [GLCore]: Loading stock shader.
> [INFO] [slang]: Building pass #0 (N/A)
> [INFO] [slang]: Reflection
> [INFO] [slang]:   Textures:
> [INFO] [slang]:      Source (#0)
> [INFO] [slang]:
> [INFO] [slang]:   Uniforms (Vertex: yes, Fragment: no):
> [INFO] [slang]:   Push Constants (Vertex: no, Fragment: no):
> [INFO] [slang]:      MVP (Offset: 0)
> [INFO] [slang]:
> [INFO] [slang]:   Parameters:
> [INFO] [GLCore]: Not using frame history.
> [INFO] [GLCore]: Not using framebuffer feedback.
> [INFO] [Font]: Using font rendering backend: freetype.
> [INFO] [Video]: Found display server: win32
> [DEBUG] video_driver_get_flags() = 6256, GFX_CTX_FLAGS_SHADERS_SLANG set!
> [DEBUG] video_context_driver_get_flags() = 129

At this point `video_driver_get_flags()` starts working and the slang shader flag is set, which is why we are then able to use shaders.

In vulkan, the Slang shader flag is set from the beginning (and when video is set up):
> [DEBUG] video_context_driver_get_flags() = 2049, GFX_CTX_FLAGS_SHADERS_SLANG set!
> (...)
> [DEBUG] video_driver_get_flags() = 6248, GFX_CTX_FLAGS_SHADERS_SLANG set!

I don't fully understand the concept of hardware contexts, but this looks like waiting for video to be set up before trying to compile shaders would work around the problem (and that's probably what's happening when starting from the menu).

Apart from that the question is why `video_context_driver_get_flags()` doesn't have the Slang shader flag set.
This is as far as I've come yet.

**Comment by LazyBumHorse** [comment]
Okay, that took forever, but I think I finally found it - and I'm still confused.

First of all, from what I gathered, those flags apparently aren't actually polled from the video driver, the OS or anything similar, instead they are hardcoded.

`video_driver_get_flags()` gets its flags from `current_video->poke_interface` which for glcore is `gl_core_get_poke_interface`, which uses `gl_core_get_flags`, which is this:
https://github.com/libretro/RetroArch/blob/5c7a5fdba0120566519bc85e42640fb2804256bb/gfx/drivers/gl_core.c#L1750-L1761
Adding these flags together gets 6256, which you see in the log above.

Now for the confusing part, `wgl_ctx`:
https://github.com/libretro/RetroArch/blob/5c7a5fdba0120566519bc85e42640fb2804256bb/gfx/drivers_context/wgl_ctx.c#L805-L829
Especially this:
https://github.com/libretro/RetroArch/blob/5c7a5fdba0120566519bc85e42640fb2804256bb/gfx/drivers_context/wgl_ctx.c#L819
In other context drivers like `wayland_ctx` this looks like this
https://github.com/libretro/RetroArch/blob/5c7a5fdba0120566519bc85e42640fb2804256bb/gfx/drivers_context/wayland_ctx.c#L1974-L1982
For vulkan (wgl and wayland) it also looks like this.
So I can only assume this to be an oversight.

I guess this is all there is to this? 3 missing lines of code?
Quite anticlimactic, but at least easy to fix.
Pull request soon.

**Comment by LazyBumHorse** [comment]
@KrokusPokus Which systems did you test this on? Was it only broken on Windows?
Just to be sure this only applies to wgl.

**Comment by twinaphex** [comment]
@LazyBumHorse yeah if you can reliably set the 'api' variable inside the WGL context driver, that would be preferable vs. calling the video driver and grabbing the 'ident' name to make a determination on this. So indeed, this could be an oversight.

Looking forward to your pull request if it indeed fixes your issues.

**Comment by orbea** [comment]
@LazyBumHorse Is this now fixed?

**Comment by LazyBumHorse** [comment]
For wgl it should be (works for me), but maybe there are similar errors in other context drivers, I'll quickly check them and if I don't find anything I guess I'll close the issue.

**Comment by orbea** [comment]
I can still reproduce this with KMS and linux actually.

**Comment by orbea** [comment]
Should be fixed for KMS + glcore too now.

https://github.com/libretro/RetroArch/pull/8895

**Comment by KrokusPokus** [comment]
@LazyBumHorse I only tested on Windows 7 x64 on 1.7.7 stable

**Comment by LazyBumHorse** [comment]
I'm pretty sure metal (cocoa_gl_ctx) is also affected, it's pretty messy to figure out what works and what doesn't.
The underlying problem so far has always been using `video_shader_parse` too early, here's the call stack:
```
gl_core_init()
 gl_core_init_filter_chain()
  video_shader_parse_type()
   video_shader_get_type_from_ext()
    video_driver_get_all_flags()
     video_driver_get_flags() <- fails when `video_driver_poke` isn't set up yet
     video_context_driver_get_flags()
```
`video_driver_poke` is set up right _after_ `init()`:
https://github.com/libretro/RetroArch/blob/580bb2cf54ba06fd9f7de44e6fd83515fde34a5c/gfx/video_driver.c#L1069-L1084
Since it takes `video_driver_data` from `init()`, it cannot simply be moved before it.

Adding flags to the context driver to solve this seems more like a hack, but I also don't understand what a context actually is and why it is already set up this way, so...

I definitely need to take a break for today, but I thought at least leaving this here, no half-baked suggestions either.

**Comment by LazyBumHorse** [comment]
@KrokusPokus It's odd that shaders loaded properly in some cases then, I tried genesis_plus_gx_libretro before too and it did not work, same system.
You can try the nightly on the [build bot](https://buildbot.libretro.com/nightly/windows/x86_64/) tomorrow (that should be 2019-06-02_RetroArch.7z), see if you can still find any issue.

**Comment by KrokusPokus** [comment]
@LazyBumHorse I confirm that slang shader loading is fixed with me in current nightly.

**Comment by LazyBumHorse** [comment]
@KrokusPokus Great, thanks for testing!

**Comment by axredneck** [comment]
Same issue here but with WIMP GUI instead of CLI.
Arch Linux, Retroarch git master built just now, Mesen core (didn't try others yet)
[retroarch.log](https://github.com/libretro/RetroArch/files/3294805/retroarch.log)

**Comment by LazyBumHorse** [comment]
@axredneck You can try the next nightly and reopen if it's still a problem.
